File: 02_paradigm_exploration/sorting_and_searching/binary_search.scala
object BinarySearchApp {
  def binarySearch(arr: Vector[Int], x: Int): Int = {
    def search(low: Int, high: Int): Int = {
      if (low > high) -1
      else {
        val mid = (low + high) / 2
        if (arr(mid) == x) mid
        else if (arr(mid) > x) search(low, mid - 1)
        else search(mid + 1, high)
      }
    }
    search(0, arr.length - 1)
  }

  def main(args: Array[String]): Unit = {
    val arr = Vector(1, 3, 5, 7, 9, 11, 13, 15)
    val x = 7

    val start = System.nanoTime()
    val index = binarySearch(arr, x)
    val end = System.nanoTime()

    println(s"Element $x found at index: $index")
    println(f"Execution time: ${(end - start) / 1e6}%.3f ms")
  }
}

File: 02_paradigm_exploration/sorting_and_searching/binary_search.rb
class BinarySearch
    def initialize(arr)
      @arr = arr
    end
  
    def search(x)
      low = 0
      high = @arr.length - 1
  
      while low <= high
        mid = (low + high) / 2
        if @arr[mid] == x
          return mid
        elsif @arr[mid] < x
          low = mid + 1
        else
          high = mid - 1
        end
      end
      -1
    end
  end
  
  if __FILE__ == $0
    arr = [1, 3, 5, 7, 9, 11, 13, 15]
    x = 7
    bs = BinarySearch.new(arr)
  
    start = Time.now
    index = bs.search(x)
    duration = Time.now - start
  
    puts "Element #{x} found at index: #{index}"
    puts "Execution time: #{(duration * 1000).round(3)} ms"
  end
  